addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/workers/RuleEvaluationWorker.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * (C) Copyright 2019 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Guillaume RENARD
+ */
+package org.nuxeo.retention.workers;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.DocumentNotFoundException;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.api.repository.RepositoryManager;
+import org.nuxeo.ecm.core.work.AbstractWork;
+import org.nuxeo.retention.adapters.Record;
+import org.nuxeo.retention.service.RetentionManager;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Worker scheduled to proceed #{link {@link org.nuxeo.retention.service.RetentionManager#evalExpressionEventBasedRules}
+ * on document ids with events set.
+ *
+ * @since 11.1
+ */
+public class RuleEvaluationWorker extends AbstractWork {
+
+    private static final Logger log = LogManager.getLogger(Record.class);
+
+    private static final long serialVersionUID = 1L;
+
+    public static final String TITLE = "Retention Rule Evaluation Work";
+
+    public static final String CATEGORY = "retentionRuleWork";
+
+    protected Map<String, Set<String>> docsToCheckAndEvents;
+
```

Comment:
i am wondering if we can find ourself with race condition when populating this map -> perhaps should we use the `ConcurrentHashMap`

Reply:
No need, Work construction and evaluation is always single-threaded.